Sources/EnumExtensions.swift

```diff
@@ -0,0 +1,42 @@
+//
+//  EnumExtensions.swift
+//  EZSwiftExtensions
+//
+//  Created by furuyan on 2017/08/26.
+//  Copyright © 2017年 Goktug Yilmaz. All rights reserved.
+//
+
+import Foundation
+
+protocol EnumCollection {
+    associatedtype Case = Self
+}
+
+extension EnumCollection where Case: Hashable {
+    fileprivate static var iterator: AnyIterator<Case> {
+        var n = 0
+        return AnyIterator {
+            defer { n += 1 }
+
+            let next = withUnsafePointer(to: &n) {
+                UnsafeRawPointer($0).assumingMemoryBound(to: Case.self).pointee
```

Comment:
the three worst words in Programming languages. Unsafe, raw and pointer. :P

What is the expectation on calling an index that is outside of an enum specificiation (e.g. in your example case 3?).

Reply:
In Swift, `UnsafePointer` means it's just not under the ARC.
And `raw` 's lifetime is limited in  `withUnsafePointer` block.
So I think it doesn't matter.

>What is the expectation on calling an index that is outside of an enum specificiation (e.g. in your example case 3?).

It expects nil.
`let castedNext = next as? Case // castedNext is nil`

Reply:
Whats the garbage collection expectaion on the next pointer here then, if its not covered by ARC?

Reply:
According to the `withUnsafePointer`'s document, 
`The pointer argument is valid only for the duration of the function’s execution.`
And besides UnsafePointer is constructed by struct, so we don't need to think about it.